STIPS/STIP-005.md

```diff
@@ -0,0 +1,877 @@
+
+<li><em>DIM</em> calls <em>PerpV2Module</em> issuance hook
+
+<li><em>PerpV2Module</em> calls <strong><em>sync()</em></strong> inside the module issuance hook
+</li>
+</ul>
+   </td>
+  </tr>
+  <tr>
+   <td><em>sync() </em>
+<p>
+(flow summary)
+   </td>
+   <td>
+  <ul>
+  <li>Reads the virtual token balances of the virtual SetToken from PerpV2Module</li>
+  <li>Updates the default position units  on the virtual SetToken </li>
+    <ul>
+      <li>Only done if the new unit calculation does not equal existing unit </li>
+      <li>Only necessary if a liquidation has occurred. </li>
+      <li>Interest accrual is not an issue here </li>
+    </ul>
+  <li>Updates external USDC position on the real SetToken </li>
+  </ul>
+   </td>
+  </tr>
+<tr>
+   <td>sync()
+  <p>(updating virtual default positions if necessary)
+   </td>
+   <td><em>newUnitForVirtualAsset = vSetToken.balanceOf(vAsset) / realSetToken.totalSupply()</em>
+   </td>
+  </tr>
+  <tr>
+   <td>sync()
+  <p>(Update external USDC position on the real SetToken)
+   </td>
+   <td>
+     <p> First determine the additional amount of virtual assets we need exposure to: </p>
+     <p> &nbsp;&nbsp;<em> newVAssetUnit = mintQuantity * vSetToken.getDefaultPositionRealUnit(vAsset) </em> </p>
+     <p> Calculate the amount of USDC required to buy newVAssetUnit by simulating a trade on Perp’s vAMM. The delta quote returned by the simulation will include realized funding payments and fees. These will be distributed across the real SetToken's external USDC position. </p>
+     <p> &nbsp;&nbsp;<em> deltaQuote = getQuoteForVAsset(vAsset, newVAssetUnit, direction=long | short)</em> </p>
+     <p> Update the real SetToken’s external position for USDC as below. (For a SetToken with multiple positions in multiple virtual assets we would have to sum up all delta) </p>
+     <p> &nbsp;&nbsp;<em> newTotalSupply = realSetToken.totalSupply + mintQuantity </em> </p>
+     <p> &nbsp;&nbsp;<em> newUSDCTotal = externalUSDCPositionUnit * realSetToken.totalSupply + deltaQuote </em> </p>
+           <p> &nbsp;&nbsp;<em> newExternalPositionUnit = newUSDCTotal / newTotalSupply </em> </p>
+   </td>
+  </tr>
+  <tr>
+   <td>DIM: transfer USDC
+   </td>
+   <td>
+    <p>DIM reads external USDC position updated during <em>sync()</em>  from real setToken and transfers in USDC amount </p>
+    <p> <em> &nbsp;&nbsp; usdcAmount = mintQuantity * external position on real SetToken</em> </p>
+   </td>
+  </tr>
+  <tr>
+   <td>Component IssuanceHook flow
+   </td>
+   <td>First, Invoke SetToken to deposit the received USDC from issuer to PERP vault
+<p>
+Then, iterate on default positions of virtual SetToken, opening positions
+   <p> &nbsp;&nbsp;<em>for asset[i], positionUnit[i] of virtual SetToken</em></p>
+     <p> &nbsp;&nbsp;&nbsp;&nbsp;<em>  tradeAmount = positionUnit[i] * mintQuantity</em> </p>
+     <p> &nbsp;&nbsp;&nbsp;&nbsp;<em>  ClearingHouse.openPositon(asset[i], tradeAmount)</em> </p>
```

Comment:
So we would be trading a fixed amount of vUSDC for some amount of vETH? How does this affect our leverage ratios?

Reply:
The `vETH` amount is fixed here which is stored in the `tradeAmount` variable. Our leverage ratios wouldn't be affected, because we are trying to replicate the current leverage ratio in our calculations above. On a very high level, the idea is:

1. opening a 2x long position
```
Let, 1 ETH = 100 USDC
Deposit, 100 USDC, mint 200 vUSDC, and swap for 2 vETH.

Our final balances on perp v2
in Vault : 100 USDC
in ClearingHouse : 2 vETH

Mint 1 SetToken  
SetToken total supply = 1 * 10^18
Final position units
- Virtual SetToken
  - vETH Position Unit = 2 * 10^18
- SetToken
  - external USDC position = 100 * 10 ^ 6
```
2. Adding to the 2x long position (which is what we do during issuance)
```
Let mintQuantity = 1 * 10^18 SetToken
First, We sync our virtual positions
Then, tradeAmount = preciseMul(2 * 10^18, 1 * 10^18) = 2 * 10^18 or in words, 2 vETH
In the above steps, we determine the amount of USDC that needs to be passed in, which would be 100 USDC
Then CH.openPosition(vETH, tradeAmount)

Our final balances on perp v2
in Vault : 200 USDC
in ClearingHouse : 4 vETH


Mint 1 SetToken  
SetToken total supply = 2 * 10^18
Final position units
- Virtual SetToken
  - vETH Position Unit = 2 * 10^18
- SetToken
  - external USDC position = 100 * 10 ^ 6
```